sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -3928,3 +3928,128 @@ case class ArrayExcept(left: Expression, right: Expression) extends ArrayBinaryL
 
   override def prettyName: String = "array_except"
 }
+
+@ExpressionDescription(
+  usage = """
+  _FUNC_(array, element) - Returns an array of appending an element to the end of an array
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_(array(1, 2, 3), 3);
+       [1,2,3,3]
+      > SELECT _FUNC_(array(1, 2, 3), null);
+       [1,2,3,null]
+      > SELECT _FUNC_(a, e) FROM VALUES (array(1,2), 3), (array(3, 4), null), (null, 3) tbl(a, e);
+       NULL
+       [1,2,3]
+       [3,4,null]
+  """,
+  since = "3.0.0")
+case class ArrayAppend(left: Expression, right: Expression) extends BinaryExpression {
```

Comment:
This is equivalent to Concat(Seq(left + CreateArray(Seq(right)))). Shall we use existing expressions instead adding duplicate codes?

Reply:
There are some cases result different.

Reply:
Besides the Concat function for array is completely different from Postgres, I think that we have it implemented wrong

Reply:
Could you please put the case here?

Reply:
```sql
postgres=# select array_cat(array[1,2], array[2]);
 array_cat
-----------
 {1,2,2}
(1 row)

postgres=# select concat(array[1,2], array[2]);
  concat
----------
 {1,2}{2}
(1 row)
```

```sql
select concat(array(1,2), array(2));
[1,2,2]
```

Reply:
Some bugs, also
```sql
spark-sql> select concat(null, array(1));
Error in query: cannot resolve 'concat(CAST(NULL AS STRING), array(1))' due to data type mismatch: input to function concat should all be the same type, but it's [string, array<int>]; line 1 pos 7;
'Project [unresolvedalias(concat(cast(null as string), array(1)), None)]
+- OneRowRelation

spark-sql>
```

```
spark-sql> select concat(array(1), array(null));
19/11/04 10:26:33 ERROR SparkSQLDriver: Failed in [select concat(array(1), array(null))]
scala.MatchError: NullType (of class org.apache.spark.sql.types.NullType$)
	at org.apache.spark.sql.catalyst.expressions.Cast.castToInt(Cast.scala:434)
	at org.apache.spark.sql.catalyst.expressions.Cast.org$apache$spark$sql$catalyst$expressions$Cast$$cast(Cast.scala:622)
	at org.apache.spark.sql.catalyst.expressions.Cast.castArray(Cast.scala:563)
	at org.apache.spark.sql.catalyst.expressions.Cast.org$apache$spark$sql$catalyst$expressions$Cast$$cast(Cast.scala:627)
	at org.apache.spark.sql.catalyst.expressions.Cast.cast$lzycompute(Cast.scala:639)
```

Reply:
```
spark-sql> select concat(null, array(1));
```
Is this valid in `ArrayAppend`? I think `checkInputDataTypes` just throws an exception in the case.

```
spark-sql> select concat(array(1), array(null));
```
What's the spark version that you used? I tried in the current master;
```
scala> sql("""select concat(array(1), array(null))""").show()
+-----------------------------+
|concat(array(1), array(NULL))|
+-----------------------------+
|                         [1,]|
+-----------------------------+
```

Anyway, I think you just need to handle these error cases for ArrayAppend.

Reply:
> ```sql
> postgres=# select array_cat(array[1,2], array[2]);
>  array_cat
> -----------
>  {1,2,2}
> (1 row)
> 
> postgres=# select concat(array[1,2], array[2]);
>   concat
> ----------
>  {1,2}{2}
> (1 row)
> ```
> 
> ```sql
> select concat(array(1,2), array(2));
> [1,2,2]
> ```

This looks like Spark concat for array type is like Presto's array_cat.

Reply:
```
-- !query 13
select array_append(array(1,2), null)
-- !query 13 schema
struct<array_append(array(1, 2), NULL):array<int>>
-- !query 13 output
[1,2,null]
```

Reply:
Sorry I made a mistake of the cases, i will recheck the master branch

Reply:
I just have a quick research on the master branch which I build on 2019-10-29, 

spark-sql> select concat(array(null), null);
Error in query: cannot resolve 'concat(array(NULL), CAST(NULL AS STRING))' due to data type mismatch: input to function concat should all be the same type, but it's [array<null>, string]; line 1 pos 7;
'Project [unresolvedalias(concat(array(null), cast(null as string)), None)]
+- OneRowRelation

spark-sql> select concat(array(1, 2), null);
Error in query: cannot resolve 'concat(array(1, 2), CAST(NULL AS STRING))' due to data type mismatch: input to function concat should all be the same type, but it's [array<int>, string]; line 1 pos 7;
'Project [unresolvedalias(concat(array(1, 2), cast(null as string)), None)]
+- OneRowRelation

spark-sql> select concat(array(1, 2), array(null));
[1,2,null]
Time taken: 2.908 seconds, Fetched 1 row(s)
```

Also, currently I don't know how to change `select array_append(null, 3);` this to concat

Reply:
> > ```sql
> > postgres=# select array_cat(array[1,2], array[2]);
> >  array_cat
> > -----------
> >  {1,2,2}
> > (1 row)
> > 
> > postgres=# select concat(array[1,2], array[2]);
> >   concat
> > ----------
> >  {1,2}{2}
> > (1 row)
> > ```
> > 
> > 
> > ```sql
> > select concat(array(1,2), array(2));
> > [1,2,2]
> > ```
> 
> This looks like Spark concat for array type is like Presto's array_cat. May not means it is wrong.

Concat used to be a string function as Hive before 2.4, and I do not know which standard or practice it follows now

Reply:
Hi @maropu , the key difference is when the array is null and the new element is not, array_append will create new array but concat's result will be null. please see the following cases.
```
-- !query 18
select array_append(a, e) from VALUES (array(1,2), 3), (array(3, 4), null), (null, 3), (null, null) tbl(a, e)
-- !query 18 schema
struct<array_append(a, e):array<int>>
-- !query 18 output
NULL
[1,2,3]
[3,4,null]
[3]

-- !query 19
select concat(a, array(e)) from VALUES (array(1,2), 3), (array(3, 4), null), (null, 3), (null, null) tbl(a, e)
-- !query 19 schema
struct<concat(a, array(e)):array<int>>
-- !query 19 output
NULL
NULL
[1,2,3]
[3,4,null]
```

Reply:
This is what postgres results：
```SQL
postgres=# select array_append(a, e) from (VALUES (array[1,2], 3), (array[3, 4], null), (null, 3), (null, null)) tbl(a, e);
 array_append
--------------
 {1,2,3}
 {3,4,NULL}
 {3}
 {NULL}
(4 rows)
```

Reply:
But, I personally think it is more natural that the result is null if the left input (array) is null. WDYT? @viirya @cloud-fan

Reply:
yea, I think null is not an empty array, so appending something to a null (left) sounds to be a null too.

Reply:
also cc @HyukjinKwon @gatorsmile @wangyum @dongjoon-hyun

Reply:
I don't have a strong preference. Seems fine to follow postgres here, i.e. `array_append` is not nullable.